bundles/org.openhab.binding.openthermgateway/src/main/java/org/openhab/binding/openthermgateway/internal/OpenThermGatewaySocketConnector.java

```diff
@@ -96,51 +90,49 @@ public void run() {
                     if (message != null) {
                         handleMessage(message);
                     } else {
-                        logger.debug("Connection closed by OpenTherm Gateway");
+                        logger.debug("Received NULL message from OpenTherm Gateway (EOF)");
                         break;
                     }
                 }
-
-                logger.debug("Stopping OpenThermGatewaySocketConnector");
             } finally {
-                connected = false;
-
                 logger.debug("OpenThermGatewaySocketConnector disconnected");
-                callback.disconnected();
+                callback.connectionStateChanged(ConnectionState.DISCONNECTED);
             }
         } catch (IOException ex) {
             logger.warn("Unable to connect to the OpenTherm Gateway.", ex);
-            callback.disconnected();
+            callback.connectionStateChanged(ConnectionState.DISCONNECTED);
         }
     }
 
     @Override
     public void stop() {
         logger.debug("Stopping OpenThermGatewaySocketConnector");
         stopping = true;
+        Thread.currentThread().interrupt();
     }
 
     @Override
     public boolean isConnected() {
-        return connected;
+        return (socket != null && socket.isConnected());
     }
 
     @Override
     public void sendCommand(GatewayCommand command) {
```

Comment:
The method sendCommand() can be called both from outside in the handler, and from inside the `run()` method, which are on different threads, so there is a risk of clashes; therefore it needs to be `synchronized`..

```
@Override
    public synchronized void sendCommand(GatewayCommand command) {```